ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,520 @@
+	for lastQueueEntry := &tasksQueue; ; lastQueueEntry = lastQueueEntry.next {
+		if lastQueueEntry.next == nil {
+			tasksCount = int64(lastQueueEntry.baseIdx + lastQueueEntry.used)
+			break
+		}
+	}
+
+	// update all the groups task :
+	// allocate the correct number of balances, as well as
+	// enough space on the "done" channel.
+	allBalances := make([]loadedAccountDataEntry, totalBalances)
+	allResources := make([]loadedResourcesEntry, totalResources)
+	usedBalances := 0
+	usedResources := 0
+
+	groupDoneCh := make(chan int, len(groupsReady))
+	const dependencyFreeGroup = -int64(^uint64(0)/2) - 1
+	for grpIdx := range groupsReady {
+		gr := &groupsReady[grpIdx]
+		gr.groupTaskIndex = grpIdx
+		gr.incompleteCount = int64(gr.balancesCount + gr.resourcesCount)
+		gr.balances = allBalances[usedBalances : usedBalances+gr.balancesCount]
+		if gr.resourcesCount > 0 {
+			gr.resources = allResources[usedResources : usedResources+gr.resourcesCount]
+			usedResources += gr.resourcesCount
+		}
+		usedBalances += gr.balancesCount
+		if gr.incompleteCount == 0 {
+			gr.incompleteCount = dependencyFreeGroup
+		}
+	}
+
+	taskIdx := int64(-1)
+	defer atomic.StoreInt64(&taskIdx, tasksCount)
+	// create few go-routines to load asyncroniously the account data.
+	for i := 0; i < asyncAccountLoadingThreadCount; i++ {
+		go p.asyncPrefetchRoutine(&tasksQueue, &taskIdx, groupDoneCh)
+	}
+
+	// iterate on the transaction groups tasks. This array retains the original order.
+	completed := make(map[int]bool)
+	for i := 0; i < len(p.groups); {
+	wait:
+		if incompleteCount := atomic.LoadInt64(&groupsReady[i].incompleteCount); incompleteCount > 0 || (incompleteCount != dependencyFreeGroup && !completed[i]) {
+			select {
+			case doneIdx := <-groupDoneCh:
+				if doneIdx < 0 {
+					doneIdx = -doneIdx
+					// if there is an error, report the error to the output channel.
+					p.outChan <- loadedTransactionGroup{
+						err: groupsReady[doneIdx].err,
+					}
+					return
+				}
+				if doneIdx > i {
+					// mark future txn as ready.
+					completed[doneIdx] = true
+					goto wait
+				} else if doneIdx < i {
+					// it was already processed.
+					goto wait
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+		next := i
+		for ; next < len(p.groups); next++ {
+			if !completed[next] {
+				if next > i {
+					i = next
+					goto wait
+				}
+				// next == i
+			}
+
+			delete(completed, next)
+
+			// if we had no error, write the result to the output channel.
+			// this write will not block since we preallocated enough space on the channel.
+			p.outChan <- loadedTransactionGroup{
+				accounts:  groupsReady[next].balances,
+				resources: groupsReady[next].resources,
+			}
+		}
+		// if we get to this point, it means that we have no more transaction to process.
+		break
+	}
+}
+
+func (gt *groupTask) markCompletionAcct(idx int, br loadedAccountDataEntry, groupDoneCh chan int) {
+	gt.balances[idx] = br
+	if 0 == atomic.AddInt64(&gt.incompleteCount, -1) {
```

Comment:
vscode linter complains about yoda expressions

Reply:
I see not what's wrong with Yoda expressions, but mind not swapping thee.